staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go

```diff
@@ -0,0 +1,617 @@
+			return cached, fmt.Errorf("failed to download discovery for %s: %v", path, writer.String())
+		}
+
+		parsed := &metav1.APIResourceList{}
+		if err := runtime.DecodeInto(scheme.Codecs.UniversalDecoder(), writer.data, parsed); err != nil {
+			return cached, err
+		}
+
+		// Create a discomap with single group-version
+		resources, err := endpoints.ConvertGroupVersionIntoToDiscovery(parsed.APIResources)
+		if err != nil {
+			return cached, err
+		}
+
+		discoMap := map[metav1.GroupVersion]apidiscoveryv2beta1.APIVersionDiscovery{
+			// Convert old-style APIGroupList to new information
+			gv: {
+				Version:   gv.Version,
+				Resources: resources,
+			},
+		}
+
+		cached = cachedResult{
+			discovery:   discoMap,
+			lastUpdated: now,
+		}
+
+		// Don't bother saving result, there is no ETag support on this endpoint
+		dm.cachedResults[info.service] = cached
+
+		return cached, nil
+
+	case http.StatusOK:
+		parsed := &apidiscoveryv2beta1.APIGroupDiscoveryList{}
+		if err := runtime.DecodeInto(scheme.Codecs.UniversalDecoder(), writer.data, parsed); err != nil {
+			return cached, err
+		}
+		klog.Infof("DiscoveryManager: Successfully downloaded discovery for %s", info.service.String())
+
+		// Convert discovery info into a map for convenient lookup later
+		discoMap := map[metav1.GroupVersion]apidiscoveryv2beta1.APIVersionDiscovery{}
+		for _, g := range parsed.Items {
+			for _, v := range g.Versions {
+				discoMap[metav1.GroupVersion{Group: g.Name, Version: v.Version}] = v
+			}
+		}
+
+		dm.resultsLock.Lock()
+		defer dm.resultsLock.Unlock()
+
+		// Save cached result
+		cached = cachedResult{
+			discovery:   discoMap,
+			etag:        writer.Header().Get("Etag"),
+			lastUpdated: now,
+		}
+		dm.cachedResults[info.service] = cached
+		return cached, nil
+
+	default:
+		dm.resultsLock.Lock()
+		defer dm.resultsLock.Unlock()
+
+		// Unhandled response. Mark information as stale.
+		// Try again later.
+		//!TODO: After a few tries, just wipe it out?
+		// 	or after certain time?
+		if !cached.stale {
+			cached = cachedResult{
+				discovery:   cached.discovery,
+				etag:        cached.etag,
+				lastUpdated: now,
+				stale:       true,
+			}
+			dm.cachedResults[info.service] = cached
+		}
+
+		klog.Infof("DiscoveryManager: Failed to download discovery for %v: %v %s",
+			info.service.String(), writer.respCode, writer.data)
+		return cached, fmt.Errorf("service %s returned non-success response code: %v",
+			info.service.String(), writer.respCode)
+	}
+}
+
+// Try to sync a single APIService.
+func (dm *discoveryManager) syncAPIService(apiServiceName string) error {
+	dm.servicesLock.RLock()
+	info, exists := dm.apiServices[apiServiceName]
+	dm.servicesLock.RUnlock()
+
+	gv := helper.APIServiceNameToGroupVersion(apiServiceName)
+	mgv := metav1.GroupVersion{Group: gv.Group, Version: gv.Version}
+
+	if !exists {
+		// apiservice was removed. remove it from merged discovery
+		dm.mergedDiscoveryHandler.RemoveGroupVersion(mgv)
```

Comment:
this creates an interesting race when creating CRDs to replace apiservices and removing them.   That's rare enough I think I can live with the risk if it will settle back out quickly.  If I have flow like

1. create apiservice for foo/v1
2. create crd bar in foo/v1
3. delete apiservice foo/v1

how long does it take for discovery to re-show foo/v1?

Reply:
What's the answer to this question?

Reply:
The crd registration_controller [is responsible](https://github.com/kubernetes/kubernetes/blob/c20ab846924827cbe356a763a6c40cf607beee3f/pkg/controlplane/controller/crdregistration/crdregistration_controller.go#L209-L218) for reconciling CRDs and creating APIServices based on them. It uses [`AddEventHandler`](https://github.com/kubernetes/kubernetes/blob/c20ab846924827cbe356a763a6c40cf607beee3f/pkg/controlplane/controller/crdregistration/crdregistration_controller.go#L74-L101) to watch the CRDs informers. This uses the default resync period from the factory, which apiextensions-apiserver has configured to be [5 minutes](https://github.com/kubernetes/kubernetes/blob/6c3d6c46309d96f6b3b64d8441a2c67998cadb2b/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go#L176).

I think at a worst case that we would have to wait for a full resync 5 minutes later to have the CRD resynced for its APIService to be recreated. This is the same behavior for today's discovery endpoints

Reply:
Five minutes works ok for me for this edge case.